sdk/core/Azure.Core/src/DynamicData/AllowListConverterFactory.cs

```diff
@@ -0,0 +1,196 @@
+    {
+        public static readonly AllowListConverterFactory Default = new();
+
+        public override bool CanConvert(Type typeToConvert)
+        {
+            return !IsAllowedType(typeToConvert);
+        }
+
+        public static bool IsAllowedType(Type type)
+        {
+            if (IsAllowedKnownType(type))
+            {
+                return true;
+            }
+
+            return IsAllowedPoco(type, new List<Type>());
+        }
+
+        private static bool IsAllowedPoco(Type type, List<Type> ancestorTypes)
+        {
+            if (!HasPublicParameterlessConstructor(type) && !IsAnonymousType(type))
+            {
+                return false;
+            }
+
+            foreach (PropertyInfo property in type.GetProperties())
+            {
+                if (!property.CanRead)
+                {
+                    return false;
+                }
+
+                if (!property.CanWrite && !IsAnonymousType(type))
+                {
+                    return false;
+                }
+
+                if (IsAllowedKnownType(property.PropertyType))
+                {
+                    continue;
+                }
+
+                // Trust but verify
+                if (ancestorTypes.Contains(property.PropertyType))
+                {
+                    continue;
+                }
+
+                // Recurse
+                ancestorTypes.Add(type);
+                if (!IsAllowedPoco(property.PropertyType, ancestorTypes))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private static bool HasPublicParameterlessConstructor(Type type)
+        {
+            return type.GetConstructor(Type.EmptyTypes) != null;
+        }
+
+        private static bool IsAnonymousType(Type type)
+        {
+            return
+                type.Namespace == null &&
+                Attribute.IsDefined(type, typeof(CompilerGeneratedAttribute), false);
+        }
+
+        private static bool IsAllowedKnownType(Type type)
+        {
+            return IsAllowedPrimitive(type) ||
+                IsAllowedArray(type) ||
+                IsAllowedCollection(type) ||
+                IsAllowedInterface(type) ||
+
+                // TODO: separate out non-primitive values?
+                type == typeof(JsonElement) ||
+                type == typeof(JsonDocument) ||
+                type == typeof(MutableJsonDocument) ||
+                type == typeof(MutableJsonElement) ||
+
+                // TODO: support extensible enums?
+                type == typeof(ETag) ||
+
+                // TODO: We'll want to remove this dependency
+                type == typeof(DynamicData) ||
+
+                // TODO: Keep this?
+                type == typeof(object[])
+                ;
+        }
+
+        private static bool IsAllowedPrimitive(Type type)
+        {
+            return
+                type == typeof(bool) ||
```

Comment:
Likely faster to check `type.IsPrimitive` for most of these, then fall back to our own type checks e.g. `DateTimeOffset`.